### PR TITLE
Generalized output from SystemMonitor and modified ExtAnalyze

### PR DESCRIPTION
**Pull request by jkrajniak**
- generalized SystemMonitorOutput
- added SystemMonitorOutputCSV : .csv output
- added SystemMonitorOutputDummy : print only values on the screen
- added support for the vector values from the Observable
- minor fix in System.py
- ExtAnalyze calculates properties exactly every n*interval simulation steps

### CODE REVIEW

**Comment by MrTheodor** [comment]
Completely do not know why LBDOutput is broken. It is related to ExtAnalyze. Basically, the old approach uses `counter`, the properties were calculated at simulation steps: 1, interval+1, 2*interval+1, ...

**Comment by junghans** [comment]
@stuehn do we still need this?

**Comment by jkrajniak** [comment]
I'll try to adjust this PR to the recent changes.

File: src/analysis/Observable.hpp [hunk 1]
```diff
@@ -69,7 +69,7 @@ class Observable : public SystemAccess
     /** returns observable of type int, used for Python and on C++ level*/
     virtual int compute_int() const { return 0; };
     /** computes vector of real values (e.g. pressure tensor, ...), used on C++ level only */
-    virtual void compute_real_vector() { return; };
+    virtual std::vector<real> compute_real_vector() { return std::vector<real>(); };
```
Comment (XzzX) [inline src/analysis/Observable.hpp:72]: Is throwing a not implemented error a better default?

File: testsuite/lattice_boltzmann/test_LBMDcoupling.py [hunk 2]
```diff
@@ -177,11 +177,11 @@ def test_lbmdcoupling(self):
...
         print("total LB-MD mom:  ", ("{:>18.1e}"*3).format(*tot_mom), "\n")
-
-        # momentum checks
+    #
+    #     # momentum checks
         self.assertAlmostEqual(tot_mom[0], 0., places=10)
-        self.assertAlmostEqual(tot_mom[1], 0., places=10)
-        self.assertAlmostEqual(tot_mom[2], 0., places=10)
+    #     self.assertAlmostEqual(tot_mom[1], 0., places=10)
+    #     self.assertAlmostEqual(tot_mom[2], 0., places=10)
```
Comment (XzzX) [inline testsuite/lattice_boltzmann/test_LBMDcoupling.py:184]: Why are these commented out? Either remove or make them work?

File: testsuite/lattice_boltzmann/test_LBMDcoupling.py [hunk 3]
```diff
@@ -209,9 +209,9 @@ def test_restartlbmd(self):
         print("total LB-MD mom:  ", ("{:>18.1e}"*3).format(*tot_mom), "\n")
 
         # momentum checks
-        self.assertAlmostEqual(tot_mom[0], 0., places=10)
-        self.assertAlmostEqual(tot_mom[1], 0., places=10)
-        self.assertAlmostEqual(tot_mom[2], 0., places=10)
+        # self.assertAlmostEqual(tot_mom[0], 0., places=10)
```
Comment (XzzX) [inline testsuite/lattice_boltzmann/test_LBMDcoupling.py:212]: Why are these commented out? Either remove or make them work?

File: src/analysis/SystemMonitor.hpp [hunk 4]
```diff
@@ -35,15 +35,44 @@
...
+/**
+ * Base class for SystemMonitor output formats.
+ */
+class SystemMonitorOutput
 {
     friend class SystemMonitor;
 
+public:
+    virtual void write() = 0;
+    void setSystem(shared_ptr<System> system) { system_ = system; }
```
Comment (XzzX) [inline src/analysis/SystemMonitor.hpp:53]: Not really important, more matter of style. This will introduce an unnecessary copy, you could use const&.

**Review by XzzX — APPROVED** [review]

File: testsuite/lattice_boltzmann/test_LBMDcoupling.py (hunk 3 above)
Comment (jkrajniak) [reply testsuite/lattice_boltzmann/test_LBMDcoupling.py]: Thanks, reverted

File: testsuite/lattice_boltzmann/test_LBMDcoupling.py (hunk 2 above)
Comment (jkrajniak) [reply testsuite/lattice_boltzmann/test_LBMDcoupling.py]: Thanks, reverted

File: src/analysis/Observable.hpp (hunk 1 above)
Comment (jkrajniak) [reply src/analysis/Observable.hpp:72]: I would keep it as it is - other virtual method of this class returns already default values. I'm not sure what will break if I throw an exception here.

File: src/analysis/SystemMonitor.hpp (hunk 4 above)
Comment (jkrajniak) [reply src/analysis/SystemMonitor.hpp:53]: I think here we should copy the shared_ptr to share the ownership.

Comment (XzzX) [reply src/analysis/SystemMonitor.hpp:53]: Yes, of course, keep the shared_ptr. I meant you are copying the shared_ptr twice. Not relevant for a shared_ptr but relevant if you copy something like std::vector or std::string.

Comment (jkrajniak) [reply src/analysis/SystemMonitor.hpp:53]: So it's better to use:
```suggestion
    void setSystem(const shared_ptr<System> &system) { system_ = system; }
```
?

**Pull request merged**